InnerBlocks: introduce prop to specify render callback for each block.

ZebulanStanphill opened the pull request:
## Description
This PR adds an `__experimentalItemWrapper` prop to `InnerBlocks` and `InnerBlocks.Content`, allowing blocks to specify a component to wrap each block. The feature is designed to help solve issues discussed in #23915 and #24039.

**Update 2020-08-05:** this PR has been revised to use a more flexible callback-based API. This allows you to not only wrap all child blocks with multiple levels of parent elements, but also append stuff before and after each of them.

```jsx
<InnerBlocks
	__experimentalItemCallback={ ( item ) => (
		<>
			<section>{ item }</section>
			<p>You can do cool/weird stuff like this now.</p>
		</>
	) }
/>
```

To pull this off, it also ends up modifying `BlockList`, `BlockListBlock`, `BlockContentProvider`, and `serializeBlock`. The changes should have no effect on anything that doesn't use the new feature.

[Reasoning behind my decision to take this approach.](https://github.com/WordPress/gutenberg/issues/23915#issuecomment-664721534)

## How has this been tested?
It has not yet been tested. Currently planning on testing this via #24039.

## Todo
- [ ] Get some technical feedback to make sure the implementation is sane.
- [x] Add React Native edit implementation.
- [ ] Make sure the React Native implementation actually works.
- [ ] Add tests.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ZebulanStanphill committed:
InnerBlocks: introduce prop to specify wrapper element for children.

Initial web edit implementation. (No save or native implementations yet.)

ZebulanStanphill committed:
Initial save implementation.

noisysocks commented:
Thanks for exploring this!

Could you make something (e.g. the Navigation block) use this new API so that we can test how this approach affects things like drag and drop? I suspect there may be implications there.

I'm hesitant about how many _if specified, then render something_ props there are on `InnerBlocks`. It makes me think that we're using the wrong abstraction. For example, maybe if we made `InnerBlocks` a higher order component (or a hook?) we could lean on JSX to do this work for us.

```jsx
// This is naive as it doesn't handle `key` and `ref` properly.
<InnerBlocks>
	{ ( blocks ) => (
		<ul>
			{ blocks.map( ( block ) => ( <li>{ block }</li> ) ) }
			<li><InnerBlocks.Appender /></li>
		</ul>
	) }
</InnerBlocks>
```

ZebulanStanphill committed:
For testing purposes: wrap each child of Group in section tag.

ZebulanStanphill commented:
@noisysocks For testing purposes, I have now modified the Group block to wrap each of its children in a `<section>`. (I thought about modifying the Navigation Link block, but that one has a lot more special behavior going on that would complicate testing.) I'm building a plugin zip right now to test on one of my websites.

Your HOC idea looks pretty interesting. I'm not sure how to actually implement that concept, but it looks promising, so if you want to spin up a PR using that approach, please do so. I think something like that may be the right way to go.

ZebulanStanphill committed:
Fix save implementation over-escaping.

ZebulanStanphill commented:
I tested it, found some bugs, and fixed most of them.

Here's what works:
- The save implementation works.
- The parent selector button works.
- The up/down movers work for the nested blocks (both single and multi-selection).

Here's what still doesn't:
- Drag-and-drop into the parent between its children is not possible, though you can drop into the start of the Group block.
- The native edit implementation still hasn't even been attempted yet.
- No tests for this new API have been created.

Additionally, the only failing tests right now are caused by the temporary testing change made to the Group block's markup, so that's a good sign.

Notably, the up/down movers were already working prior to https://github.com/WordPress/gutenberg/pull/24232/commits/490dbcf3a03922692e9c978d15a0bc0cf347ba8f, but the animations were not. In fixing this, I ended up moving some of my wrapper logic from `BlockList` to `BlockListBlock`, where the block alignment wrapper is already handled. I figure that's probably where that logic belonged in the first place.

It should be noted that there is at least one major difference between the inner block item wrapper and the alignment wrapper: the former should not be considered part of the block, while the latter is (but should it)?

The alignment wrapper (for most blocks anyway) only exists in the editor, while the alignment classes are just applied to the block's normal root element in the save implementation. Ideally, the alignment should happen the same way in both implementations. Notably, the Image block uses a custom implementation which uses a wrapper for alignment in both the editor and the front-end.

To clarify, in my save implementation, the `InnerBlocks` child wrapper is placed outside of the child block's comment delimiters. This is because I don't consider the child wrapper to be a part of the child block itself. In contrast, I'd say an alignment wrapper _is_ part of its associated block.

The editor situation is interesting since there are no comment delimiters, and as such, the lines between where a block starts and ends are a bit less obvious.

Any thoughts/opinions on this?

As for fixing the drag-and-drop issue, I'm not sure how to fix that yet. Any suggestions would be very welcome.

ZebulanStanphill committed:
Move wrapper logic from BlockList to BlockListBlock.

Fixes block moving animations for inner blocks.

ZebulanStanphill committed:
Fix drag-and-drop for inner blocks.

ZebulanStanphill committed:
Initial native implementation.

ZebulanStanphill commented:
Alright, drag-and-drop is fixed.

I've also attempted to make this work with React Native. I have no clue if the native implementation works or how to test it (I can't even get the normal wp-env set up on my computer), so if anyone could test that for me, that would be great.

noisysocks commented:
I tested the web version locally and it works pretty well! Nice job @ZebulanStanphill.

I think I'm happy with the approach. I like that it's a simple change, and the fact that we've encountered this problem in two seperate places reassures me that it's a problem worth solving at a framework level. I'd like it if the API were more flexible, as above, but nothing stopping us from iterating on that later seeing as this is all `__experimental`.

It's a bit of a paradigm shift, so I'd like to hear opinions from more @WordPress/gutenberg-core developers, particularly @ellatrix and @youknowriad, before proceeding. I note that it's linked to in the agenda of today's editor chat.

ellatrix commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -197,6 +198,14 @@ function BlockListBlock( {
 
+	if ( Wrapper ) {
+		blockEdit = (
+			<Wrapper className="block-editor-block-list-block__wrapper">
```
Could we allow a callback instead? This seems much more future proof, e.g. if you'd like to append something or have multiple wrappers.

ZebulanStanphill replied:
Would it still be possible to apply the `block-editor-block-list-block__wrapper` CSS class if I used a callback? I'm using that in `packages/block-editor/src/components/use-block-drop-zone/index.js` to determine drop zones for drag-and-drop.

noisysocks replied:
In `getNearestBlockIndex`, do we need to check that `element` is a block? What else could it be?

noisysocks replied:
I spoke to @talldan about this and he suggested:

1. Remove [the `wp-block` check in `getNearestBlockIndex`](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/use-block-drop-zone/index.js#L53-L56) entirely.
2. Replace [this line which sets `elements`](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/use-block-drop-zone/index.js#L329) to search for `wp-block` instead of naively looking at the container's children. Care must be taken to only select `wp-block`s that  one level deep, though. Something like: `const blockElements = Array.from( element.current.querySelectorAll( '.wp-block:not(.wp-block .wp-block)' ) )`.

Be sure to test in multiple browsers since we're using a relatively advanced query in `querySelectorAll`.

Another option is to use `useSelect` to fetch the children's client IDs and turn them into DOM nodes using the [`BlockNodes` context](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-list/root-container.js#L83). This requires a bit more wrangling, though.

noisysocks replied:
> Something like: `const blockElements = Array.from( element.current.querySelectorAll( '.wp-block:not(.wp-block .wp-block)' ) )`.

Testing this in DevTools, it doesn't look like using `:not()` in this way works. Maybe something this then:

```js
const blockElements = difference(
	Array.from( element.current.querySelectorAll( '.wp-block' ) ),
	Array.from( element.current.querySelectorAll( '.wp-block .wp-block' ) )
);
```

You get the idea! 😅

talldan replied:
On looking again, it seems like `BlockNodes` only stores selected blocks, so using `querySelectorAll` makes sense:
https://github.com/WordPress/gutenberg/blob/65258a077e4899ee4933c79077108f56f8b5ce6d/packages/block-editor/src/components/block-list/block-wrapper.js#L87-L102

ZebulanStanphill replied:
@noisysocks Alright, I'm currently working on implementing this using the `_.difference` solution you suggested.

I do wonder if perhaps there's a more performance-friendly way to get the elements. Maybe something like this? (I have no idea if it's faster or slower.)

```js
function getBlockElements( node ) {
	const list = [];

	for ( const childNode of node ) {
		if ( childNode.classList.contains( 'wp-block' ) ) {
			list.push( childNode );
		} else if ( childNode.children.length > 0 ) {
			const grandchildren = getBlockElements( childNode );
			if ( grandchildren.length > 0 ) {
				list.push( ...grandchildren );
			}
		}
	}

	return list;
}

const blockElements = getBlockElements( element.current );
```

talldan replied:
@ZebulanStanphill An option for optimizing it that I've tested on https://github.com/WordPress/gutenberg/pull/23952 is to only get the list of elements when the user starts dragging, something like this:
```jsx
const blockElements = useRef();
const hasPosition = !! position;
useEffect( () => {
  if ( hasPosition ) {
    blockElements.current = difference(
	Array.from( element.current.querySelectorAll( '.wp-block' ) ),
	Array.from( element.current.querySelectorAll( ':scope .wp-block .wp-block' ) )
    );
  }
}, [ hasPosition ] );
```

It makes the assumption that the blocks won't change in any way after the user has started dragging, which I think is ok.

Though I personally wouldn't worry too much about optimizing unless you see performance issues.

ZebulanStanphill committed:
Remove need for 'block-editor-block-list-block__wrapper'.

This makes it possible to switch to a callback-based API, e.g. "itemWrapper={ ( item ) => <div>{ item }</div> }".

ZebulanStanphill committed:
Switch to callback-based API.

Edit implementation:

<InnerBlocks
    __experimentalItemCallback={
        ( item ) => <div>{ item }</div>
    }
/>

Save implementation:

<InnerBlocks.Content
    __experimentalItemCallback={
        ( item ) => <div>{ item }</div>
    }
/>

ZebulanStanphill commented:
I have now switched to a callback-based API, as suggested by @ellatrix. You can now do stuff like this:

```jsx
<InnerBlocks
	__experimentalItemCallback={ ( item ) => (
		<>
			<section>{ item }</section>
			<p>You can do cool/weird stuff like this now.</p>
		</>
	) }
/>
```

One thing that I've discovered while working on this PR is that the position of the drag-and-drop indicator line isn't ideal. It appears before the block it will be dropping before, but it doesn't account for any extra markup added by this callback API, so if you're applying wrappers with a lot of top padding or appending stuff before each block, the drop-indicator won't appear in the right spot.

This is not a new problem; it exists in `master` as well, though it's only made obvious when dealing with large margins between blocks. I have no idea how to solve this little visual quirk.

ZebulanStanphill commented on packages/block-editor/src/components/use-block-drop-zone/index.js:
```diff
@@ -326,7 +326,16 @@ export default function useBlockDropZone( {
 	useEffect( () => {
 		if ( position ) {
-			const blockElements = Array.from( element.current.children );
+			// Get the root elements of blocks inside the element, ignoring
+			// InnerBlocks item wrappers and the children of the blocks.
+			const blockElements = difference(
+				Array.from( element.current.querySelectorAll( '.wp-block' ) ),
+				Array.from(
+					element.current.querySelectorAll(
+						':scope .wp-block .wp-block'
+					)
+				)
+			);
```
Noting that this may not be the fastest way to do this. Here's an alternative implementation that I've considered, but I have no idea if it would be faster or slower:

```js
function getBlockElements( node ) {
	const list = [];

	for ( const childNode of node ) {
		if ( childNode.classList.contains( 'wp-block' ) ) {
			list.push( childNode );
		} else if ( childNode.children.length > 0 ) {
			const grandchildren = getBlockElements( childNode );
			if ( grandchildren.length > 0 ) {
				list.push( ...grandchildren );
			}
		}
	}

	return list;
}

const blockElements = getBlockElements( element.current );
```

ZebulanStanphill committed:
Rename __experimentalCallback to __experimentalRenderCallback.

ZebulanStanphill committed:
Apply render callback at a slightly higher level.

Fixes non-light blocks having the render callback applied inside the Block component rather than outside.

Also moves the "Edit as HTML" form of blocks inside the render callback.

ZebulanStanphill commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -93,6 +93,7 @@ function BlockListBlock( {
 	toggleSelection,
 	index,
 	enableAnimation,
+	__experimentalCallback: wrapperCallback,
```
I decided to go with `__experimentalRenderCallback`. How's that?

ZebulanStanphill commented:
I realized that non-light blocks had the render callback applied inside their `Block` component rather than outside of it, so I pushed a fix for that. This also makes it so that the render callback wraps even the "Edit as HTML" form of blocks, which is neat.

ZebulanStanphill commented:
This should be ready for review now. Remember, I need someone to test and make sure this doesn't break anything in the React Native implementation. Also, I'll remove the testing changes to the Group block when it's time to merge.

noisysocks commented:
Seems to work pretty well. Nice job.

Looks like there's one legitimate test failure buried among the failures that are to do with changing Group:

- packages/block-editor/src/components/use-block-drop-zone/test/index.js

Let's fix that up, revert the changes to Group, and then merge.

---

> Remember, I need someone to test and make sure this doesn't break anything in the React Native implementation.

@hypest: Are you able to help @ZebulanStanphill with this?

ZebulanStanphill committed:
Remove obsolete unit test.

getNearestBlockIndex now assumes all elements passed to it are the root elements of blocks; we filter out the other elements before passing the array to the function.

ZebulanStanphill commented:
@noisysocks Good catch! In this case, I had altered the code so that block root elements are determined _before_ passing them to `getNearestBlockIndex()`, so that function no longer checks that itself. Therefore, the unit test is obsolete; I've removed it now.

cameronvoell commented:
>  Remember, I need someone to test and make sure this doesn't break anything in the React Native implementation.


@noisysocks @ZebulanStanphill  I'll give this a test and review from the mobile perspective.

cameronvoell reviewed (commented):
> @noisysocks @ZebulanStanphill I'll give this a test and review from the mobile perspective.

Tested and everything looks good on the React Native side. Just to document, I ran an extended suite of automated mobile tests, which all [passed](https://app.circleci.com/pipelines/github/wordpress-mobile/gutenberg-mobile/7879/workflows/9cb75875-f8ed-4daf-a2c4-c2db3c8b0afa), and I spot checked some of our documented test cases for the mobile [Group](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md) and [Column](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/columns.md) Blocks across Android and iOS

noisysocks reviewed (approved):
This LGTM after tests pass. Remember to revert the changes to Group. Let's try it on!

ZebulanStanphill committed:
Revert temporary Group block changes.

ZebulanStanphill commented:
As a follow-up: should we replace `__experimentalAppenderTagName` with an `__experimentalAppenderCallback`? Or should we just remove `__experimentalAppenderTagName` and make the appender use the same render callback as the blocks?

ZebulanStanphill commented:
Uh-oh. While testing this feature out in #24039, I think I've discovered a major flaw in the current implementation... namely that I only made changes to serialization code, and I never changed any of the parsing code. This means that inserting a block using this API will work fine in the editor, and even saving it and viewing it on the front-end will work. But if you leave the editor and then return, it won't be able to parse the blocks using InnerBlock render callbacks, and block validation will fail..

I'm not really sure how to fix this, but it's a serious bug that renders the entire feature unusable. Any technical help would be greatly appreciated.

draganescu commented:
Howdy @ZebulanStanphill. Is this feature used anywhere currently? In the navigation editor the problem ended up being solved differently. If there is no use case perhaps the best path for this API is to be removed via a revert?

ZebulanStanphill commented:
@draganescu The feature is currently being used in #24039, as I just mentioned.

draganescu commented:
Oops! 😊

glendaviesnz commented:
@ZebulanStanphill FYI - I am not too heavily invested in this approach in #24039, so if we need to back it out for now and explore alternatives that is fine, or also happy to wait until we get the parsing side sorted.

ZebulanStanphill commented:
@glendaviesnz I still think the render callback approach is the best one, because the only other solutions I am aware of are:

- An intermediary "Gallery Item" block that's nothing more than a `<li>` that can have an Image block nested in it. This would complicate the process of adding another Image to the Gallery, and it would also complicate block navigation, unless we introduce a new API completely hide the block from the user-facing UI; and that would require progress on the PR to be halted anyway. See #7694.
- Conditional rendering of a `<li>` wrapper defined in the Image block. This feels really fragile and prone to complications with the placement of the `wp-block-image` class. It would also make it far more difficult for 3rd party image blocks to add support for being inserted into a core Gallery.

That said, it's probably best not to have a completely broken API in the code, even if it is experimental. And I am not currently aware of how to complete the implementation. So I've created #25196 to remove this API for now.

nathanharper commented:
Is there any plan to revisit this? It would be really useful for a mixed-media gallery I'm trying to build.
